cmd/search-server/main.go

```diff
@@ -0,0 +1,224 @@
+type LoggingConfig struct {
+	// Level of logging to print to standard output.
+	// Values: [ debug, info, warn, error, fatal ]
+	Level string `yaml:"level"`
+	// Format of log entries.
+	// Options: [ json, text ]
+	Format string `yaml:"format"`
+}
+
+// PubsubConfig holds pubsub (notification) configuration.
+type PubsubConfig struct {
+	// Enable Pub/Sub for event notification publishing.
+	// Values: [ true, false ]
+	Enable bool `yaml:"enable"`
+	// Project ID of the Google Cloud project to use for Pub/Sub.
+	// Reference: https://cloud.google.com/resource-manager/docs/creating-managing-projects
+	Project string `yaml:"project"`
+}
+
+// default configuration
+var config = ServerConfig{
+	Port: 8080,
+	Database: DatabaseConfig{
+		Driver: "sqlite3",
+		Config: "file:/tmp/registry.db",
+	},
+	Logging: LoggingConfig{
+		Level:  "info",
+		Format: "text",
+	},
+	Pubsub: PubsubConfig{
+		Enable:  false,
+		Project: "",
+	},
+}
+
+func main() {
+	var configPath string
+	pflag.StringVarP(&configPath, "configuration", "c", "", "The server configuration file to load.")
+	pflag.Parse()
+
+	// Use a default logger configuration until we load the server config.
+	bootLogger := log.NewLogger()
+	if configPath != "" {
+		bootLogger.Infof("Loading configuration from %s", configPath)
+		raw, err := ioutil.ReadFile(configPath)
+		if err != nil {
+			bootLogger.WithError(err).Fatal("Failed to open config file")
+		}
+		// Expand environment variables before unmarshaling.
+		expanded := []byte(os.ExpandEnv(string(raw)))
+		err = yaml.Unmarshal(expanded, &config)
+		if err != nil {
+			bootLogger.WithError(err).Fatalf("Failed to read config file")
+		}
+	}
+
+	if err := validateConfig(); err != nil {
+		bootLogger.WithError(err).Fatalf("Invalid configuration")
+	}
+
+	// Use logging options from the server config.
+	var (
+		logOpts        = loggerOptions(config.Logging)
+		logger         = log.NewLogger(logOpts...)
+		logInterceptor = interceptor.CallLogger(logOpts...)
+	)
+
+	logger.Infof("Configured port %d", config.Port)
+	listener, err := net.ListenTCP("tcp", &net.TCPAddr{
+		Port: config.Port,
+	})
+	if err != nil {
+		logger.WithError(err).Fatalf("Failed to create TCP listener")
+	}
+	defer listener.Close()
+
+	registryServer, err := registry.New(registry.Config{
+		Database:  config.Database.Driver,
+		DBConfig:  config.Database.Config,
+		LogLevel:  config.Logging.Level,
+		LogFormat: config.Logging.Format,
+		Notify:    config.Pubsub.Enable,
+		ProjectID: config.Pubsub.Project,
+	})
+	if err != nil {
+		logger.WithError(err).Fatalf("Failed to create registry server")
+	}
+
+	searchServer := search.New(search.Config{
+		Database: config.Database.Driver,
+		DBConfig: config.Database.Config,
+	}, registryServer)
+
+	grpcServer := grpc.NewServer(grpc.UnaryInterceptor(logInterceptor))
+	reflection.Register(grpcServer)
+	registry_rpc.RegisterRegistryServer(grpcServer, registryServer)
+	registry_rpc.RegisterAdminServer(grpcServer, registryServer)
```

Comment:
For convenience, the search server includes all of the methods of the main API server. This is essentially the configuration that we would have if we merged the search service handlers back into the server in the main repo.